src/pcre2_compile_class.c

```diff
@@ -1783,85 +1787,421 @@ return pptr - 1;
 
 
 
+/* ===================================================================*/
+/* Here follows a block of ECLASS-compiling functions. You may well want to
+read them from top to bottom; they are ordered from leafmost (at the top) to
+outermost parser (at the bottom of the file). */
+
+/* This function folds one operand using the negation operator.
+The new, combined chunk of stack code is written out to *pop_info. */
+
+static void
+fold_negation(eclass_op_info *pop_info, PCRE2_SIZE *lengthptr,
+  BOOL preserve_classbits)
+{
+/* If the chunk of stack code is already composed of multiple ops, we won't
+descend in and try and propagate the negation down the tree. (That would lead
+to O(n^2) compile-time, which could be exploitable with a malicious regex -
+although maybe that's not really too much of a worry in a library that offers
+an exponential-time matching function!) */
+
+if (pop_info->op_single_type == 0)
+  {
+  if (lengthptr != NULL)
+    *lengthptr += 1;
+  else
+    pop_info->code_start[pop_info->length] = ECL_NOT;
+  pop_info->length += 1;
+  }
+
+/* Otherwise, it's a nice single-op item, so we can easily fold in the negation
+without needing to produce an ECL_NOT. */
+
+else if (pop_info->op_single_type == ECL_ANY ||
+         pop_info->op_single_type == ECL_NONE)
+  {
+  pop_info->op_single_type = (pop_info->op_single_type == ECL_NONE)?
+      ECL_ANY : ECL_NONE;
+  if (lengthptr == NULL)
+    *(pop_info->code_start) = pop_info->op_single_type;
+  }
+else
+  {
+  PCRE2_ASSERT(pop_info->op_single_type == ECL_XCLASS &&
+               pop_info->length >= 1 + LINK_SIZE + 1);
+  if (lengthptr == NULL)
+    pop_info->code_start[1 + LINK_SIZE] ^= XCL_NOT;
+  }
+
+if (!preserve_classbits)
+  {
+  for (int i = 0; i < 8; i++)
+    pop_info->bits.classwords[i] = ~pop_info->bits.classwords[i];
+  }
+}
+
+
+
+/* This function folds together two operands using a binary operator.
+The new, combined chunk of stack code is written out to *lhs_op_info. */
+
+static void
+fold_binary(int op, eclass_op_info *lhs_op_info, eclass_op_info *rhs_op_info,
+  PCRE2_SIZE *lengthptr)
+{
+switch (op) {
```

Comment:
I think this is the format:
```
switch (op)
  {
  case X:
  break;
  }
```

Reply:
Thanks! I'll fix these formatting issues.